pkg/controller/infrastructure/actuator_restore.go

```diff
@@ -26,9 +27,22 @@ import (
 // Restore implements infrastructure.Actuator.
 func (a *actuator) Restore(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) error {
 	logger := a.logger.WithValues("infrastructure", client.ObjectKeyFromObject(infra), "operation", "restore")
-	terraformState, err := terraformer.UnmarshalRawState(infra.Status.State)
+
+	infraState := &State{}
+	if err := json.Unmarshal(infra.Status.State.Raw, infraState); err != nil {
+		return err
+	}
+
+	terraformState, err := terraformer.UnmarshalRawState(infraState.TerraformState)
 	if err != nil {
 		return err
 	}
+
+	patch := client.MergeFrom(infra.DeepCopy())
+	infra.Status.ProviderStatus = infraState.SavedProviderStatus
```

Comment:
What will be the consequences if `infraState.SavedProviderStatus` is nil? e.g if the cluster is restored before once reconciled with this change?

Reply:
The restoration will fail and the state will have to adapted manually to the new format. But I think this is ok since the control plane migration is still alpha.